File: module-10-fine-tuning/solutions/data-pipeline.ts
/**
 * Module 10: Data pipeline
 * Turn production logs/eval traces into training datasets.
 */
import * as fs from "fs/promises";
import type { Trial, Transcript } from "../../module-7-evals/solutions/types.js";

export interface TraceRecord {
  id: string;
  task: string;
  success: boolean;
  turns: number;
  toolCalls: string[];
  transcript: Transcript;
}

/** Extract structured trace records from eval trials */
export function extractTraces(trials: Trial[]): TraceRecord[] {
  return trials.map((trial, i) => ({
    id: `trace-${i}`,
    task: trial.transcript.task,
    success: trial.grade.passed,
    turns: trial.transcript.turns.length,
    toolCalls: trial.transcript.turns
      .flatMap((t) => t.toolCalls || [])
      .map((tc) => tc.name),
    transcript: trial.transcript,
  }));
}

/** Simple embedding-free clustering by task similarity (keyword overlap) */
export function clusterTraces(traces: TraceRecord[]): Map<string, TraceRecord[]> {
  const clusters = new Map<string, TraceRecord[]>();
  for (const trace of traces) {
    // Simple clustering by task keywords
    const key = trace.task.split(" ").slice(0, 3).join(" ").toLowerCase();
    if (!clusters.has(key)) clusters.set(key, []);
    clusters.get(key)!.push(trace);
  }
  return clusters;
}

/** Stratified sampling: pick representative traces from each cluster */
export function stratifiedSample(
  clusters: Map<string, TraceRecord[]>,
  samplesPerCluster: number = 5
): TraceRecord[] {
  const samples: TraceRecord[] = [];
  for (const [, traces] of clusters) {
    // Prioritize diverse outcomes (mix of success/failure)
    const successes = traces.filter((t) => t.success);
    const failures = traces.filter((t) => !t.success);
    const half = Math.ceil(samplesPerCluster / 2);
    samples.push(
      ...successes.slice(0, half),
      ...failures.slice(0, samplesPerCluster - Math.min(half, successes.length))
    );
  }
  return samples;
}

/** Save traces to JSONL file */
export async function saveTraces(traces: TraceRecord[], outputPath: string): Promise<void> {
  const lines = traces.map((t) => JSON.stringify(t)).join("\n");
  await fs.writeFile(outputPath, lines, "utf-8");
  console.log(`Saved ${traces.length} traces to ${outputPath}`);
}

File: module-10-fine-tuning/solutions/python/data_pipeline.py
"""Module 10: Data pipeline — turn eval traces into training datasets."""
import json


def extract_traces(trials: list) -> list[dict]:
    return [
        {
            "id": f"trace-{i}",
            "task": trial.transcript.task,
            "success": trial.grade.passed,
            "turns": len(trial.transcript.turns),
            "tool_calls": [
                tc["name"]
                for t in trial.transcript.turns
                for tc in (t.tool_calls or [])
            ],
            "transcript": trial.transcript,
        }
        for i, trial in enumerate(trials)
    ]


def cluster_traces(traces: list[dict]) -> dict[str, list[dict]]:
    clusters: dict[str, list[dict]] = {}
    for trace in traces:
        key = " ".join(trace["task"].split()[:3]).lower()
        clusters.setdefault(key, []).append(trace)
    return clusters


def stratified_sample(clusters: dict[str, list[dict]], samples_per_cluster: int = 5) -> list[dict]:
    samples = []
    for traces in clusters.values():
        successes = [t for t in traces if t["success"]]
        failures = [t for t in traces if not t["success"]]
        half = (samples_per_cluster + 1) // 2
        samples.extend(successes[:half])
        samples.extend(failures[:samples_per_cluster - min(half, len(successes))])
    return samples


def save_traces(traces: list[dict], output_path: str) -> None:
    with open(output_path, "w") as f:
        for trace in traces:
            record = {k: v for k, v in trace.items() if k != "transcript"}
            f.write(json.dumps(record, default=str) + "\n")
    print(f"Saved {len(traces)} traces to {output_path}")
